### PR TITLE
Fast path for validating static table HTTP/2 headers

### PR DESCRIPTION
**Pull request by JamesNK**
Incoming request headers are currently identified during validation by comparing header bytes. This can be made faster for headers with a static table index. Static table headers can also skip validation that names are all lower case.

This change will speed up validation of request headers. Once it is merged I think the next step is to add a method on `HttpRequestHeaders` collection that sets headers using the index instead of name. **Update:** Now also in this PR.

Note: There will probably need to be a reaction in dotnet/runtime to the `IHttpHeadersHandler.OnStaticIndexedHeader` overloads being used. If runtime doesn't want to use the static index then it is pretty easy to implement the methods like so:

```cs
public void OnStaticIndexedHeader(int index)
{
    ref readonly var entry = ref H2StaticTable.Get(index - 1);
    OnHeader(entry.Name, entry.Value);
}

public void OnStaticIndexedHeader(int index, ReadOnlySpan<byte> value)
{
    OnHeader(H2StaticTable.Get(index - 1).Name, value);
}

public void OnHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
{
    // ...
}
```

@geoffkizer I don't know if you are still looking at HttpClient perf but you might be able to get some perf gains in the client by comparing the index instead of bytes/strings.

### CODE REVIEW

File: src/Shared/runtime/Http2/Hpack/HPackDecoder.cs
Comment (gfoidl) [inline src/Shared/runtime/Http2/Hpack/HPackDecoder.cs:525]: Just as idea:
Separate this out in a struct, so the clearing can be done just by `= default`, and the JIT will [emit more efficient code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgMBAOwwMA+gAsY2ACYwoAZQyiBYAJKTVCANxce/YaPEAKKTNxaoGNBOkMANjEkBzDPICUAPxyiipqAHLY+DAAStjeMCbUPHyCImIMNq722I7Otu6ePv5BCkqqUABq2G4ArjFxXgmm3M1MjMQoDACyVn4crcmloRpaGDr6hgwAvAwADInJgyHlEVGx8dMMkjVubguL3EPlVbX1GzPbu/vcAL40d9Q09EykDAAi/UlmqZYM9lA1MAyN6ibCfA7cZ75AASZTUmm0egMMGMAx4z0ydgcThcMg83l8gQYsOGqzOjWuiwx+WyuVxBQJxWJcMq1Tq6wprQeyVa5jS4hBWjkylBlNazw63V64IhshFQpmqgAZtgdhhKQ8bkA==).
Codegen for individual field accesses should remain in the same machine code.

I have no feeling how hot this method is, so don't know if this it worth it.

File: src/Shared/runtime/Http2/Hpack/HPackDecoder.cs
Comment (JamesNK) [reply src/Shared/runtime/Http2/Hpack/HPackDecoder.cs:512]: At the time I was probably focused on getting it working and didn't put extra thought into it.

They're good improvements 👍

File: src/Shared/runtime/Http2/Hpack/HPackDecoder.cs
Comment (JamesNK) [reply src/Shared/runtime/Http2/Hpack/HPackDecoder.cs:525]: I want to leave low-level optimizations like that out of this PR, but you're welcome to take a look at HPackDecoder. There are probably a lot of places where small improvements (e.g. eliminating bounds checks, inlining methods)  can be made.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:1271]: It does feel bad to kill the entire connection and with it any concurrent streams because the client exceeded Kestrel's MaxRequestHeadersTotalSize which isn't communicated via the protocol. Can we set some state (maybe the  Http2StreamErrorException) that tells us to continue decoding the headers, but not store them, and then rethrow the Http2StreamErrorException after we're done decoding?

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:1271]: It is existing behavior from the regular header path. Create an issue for it.

**Comment by JamesNK** [comment]
@aspnet-hello benchmark http2

File: src/Shared/runtime/Http2/Hpack/HPackDecoder.cs
Comment (gfoidl) [reply src/Shared/runtime/Http2/Hpack/HPackDecoder.cs:525]: Sounds good. Will have a look after this PR is in.

**Comment by geoffkizer** [comment]
> @geoffkizer I don't know if you are still looking at HttpClient perf but you might be able to get some perf gains in the client by comparing the index instead of bytes/strings.

Yeah, we have an issue on this: https://github.com/dotnet/runtime/issues/1505

Definitely would be nice to do, but we punted it out of 5.0. Having this implemented in the decoder itself is nice and will make this easier to do in the future.

If you have scenarios where you think this would help, let me know and we can try to squeeze it into 5.0 -- but time is very short :)

**Comment by geoffkizer** [comment]
Per discussion with @karelz etc about this, it's too late to take this change in the runtime. So if/when you merge this to aspnetcore, we will stop pulling updates until 6.0 (which we were planning to do soon anyway).

**Comment by JamesNK** [comment]
Doesn't sync go in the other direction? i.e. runtime is the authoritative source of truth for the shared HTTP/2 source, and aspnetcore keeps in sync with it.

How can I change this file in aspnetcore without updating runtime? An option to avoid any behavior changes in runtime could be to have `#ifdef` to preserve what it currently does. Or turn off the sync check that aspnetcore does until runtime branches master for 6.0.

**Comment by geoffkizer** [comment]
I'm not clear on how the code sync works; I'll defer to @karelz and @scalablecory here...

**Comment by Tratcher** [comment]
The shared code syncs from runtime to aspnetcore. We should not allow them to diverge. If this can't go into runtime for 5.0 then we should hold it for both repos until 6.0.

**Comment by JamesNK** [comment]
😢 

@geoffkizer Are you ok with a PR for this in runtime if the changes are `#ifdef`ed out by default? Runtime will be unchanged, and ASP.NET Core can set a compiler constant to get static indexes.

**Comment by scalablecory** [comment]
Current plan is for runtime "master" to be .NET 6 on 8/17.

From our standpoint it can be merged at that point (assuming it comes with accompanying SocketsHttpHandler change to light it up) and still keep the repos synced.

**Comment by JamesNK** [comment]
Ok. It can wait until then.

**Comment by geoffkizer** [comment]
I'm fine with that but it seems a little weird... you're going to be syncing to your 5.0 branch from our 6.0 branch?

**Comment by Tratcher** [comment]
> I'm fine with that but it seems a little weird... you're going to be syncing to your 5.0 branch from our 6.0 branch?

Yeah, that's not good. If anything we'd change the sync bot to sync the new runtime 5.0 branch to aspnetcore's master branch. AspNetCore won't branch for 6.0 for months.

**Comment by geoffkizer** [comment]
If you really want this for 5.0, then I'd prefer to just get it in rather than do #ifdefs or something like that. It's a reasonably safe change. Needs to get into runtime this week, though.

@karelz @scalablecory Any objections here?

**Comment by JamesNK** [comment]
Hmmm, is sync only on master? If it is then that will be happening automatically until both repos branch for 6.0.

If syncing changes when branching happens then this would still be problematic.

**Comment by karelz** [comment]
@JamesNK if you can do the change in runtime (we're a bit busy to help this week before branching sadly :(), then I'd be ok to take it based on @geoffkizer's assessment of risk above in https://github.com/dotnet/aspnetcore/pull/24730#issuecomment-672270895

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:1277]: https://github.com/dotnet/aspnetcore/issues/17861

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:1266]: This of-by-one fixup should be happening inside H2StaticTable.Get. https://github.com/dotnet/aspnetcore/blob/fa2d86e5b6b88dd8d7ec4b6de9ca5640e64ed76f/src/Shared/runtime/Http2/Hpack/H2StaticTable.cs#L13

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:1280]: Can this be unified with OnHeader? The differences are very limited. Maybe use a nullable int for the non-static flow?

**Review by Tratcher — APPROVED** [review]
The H2StaticTable.Get off-by-one is leaking all over the code. Please fix it in this PR or a follow-up one.

**Comment by JamesNK** [comment]
> The H2StaticTable.Get off-by-one is leaking all over the code. Please fix it in this PR or a follow-up one.

It is tricky because the code is shared with runtime. Next time I update it in both places I'll make that change.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:1266]: https://github.com/dotnet/aspnetcore/pull/24730#issuecomment-673722239

**Pull request merged**
